streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -653,31 +641,37 @@ private void validateStartOnce() {
      * @throws StreamsException if the Kafka brokers have version 0.10.0.x
      */
     public synchronized void start() throws IllegalStateException, StreamsException {
-        log.debug("{} Starting Kafka Stream process.", logPrefix);
-        validateStartOnce();
-        checkBrokerVersionCompatibility();
+        log.debug("{} Starting Streams client", logPrefix);
 
-        if (globalStreamThread != null) {
-            globalStreamThread.start();
-        }
+        // first set state to RUNNING before kicking off the threads,
+        // making sure the state will always transit to RUNNING before REBALANCING
+        if (setState(State.RUNNING)) {
+            checkBrokerVersionCompatibility();
 
-        for (final StreamThread thread : threads) {
-            thread.start();
-        }
+            if (globalStreamThread != null) {
+                globalStreamThread.start();
+            }
 
-        final Long cleanupDelay = config.getLong(StreamsConfig.STATE_CLEANUP_DELAY_MS_CONFIG);
-        stateDirCleaner.scheduleAtFixedRate(new Runnable() {
-            @Override
-            public void run() {
-                synchronized (stateLock) {
+            for (final StreamThread thread : threads) {
+                thread.start();
+            }
+
+            final Long cleanupDelay = config.getLong(StreamsConfig.STATE_CLEANUP_DELAY_MS_CONFIG);
+            stateDirCleaner.scheduleAtFixedRate(new Runnable() {
+                @Override
+                public void run() {
                     if (state == State.RUNNING) {
                         stateDirectory.cleanRemovedTasks(cleanupDelay);
                     }
                 }
-            }
-        }, cleanupDelay, cleanupDelay, TimeUnit.MILLISECONDS);
+            }, cleanupDelay, cleanupDelay, TimeUnit.MILLISECONDS);
 
-        log.info("{} Started Kafka Stream process", logPrefix);
+            log.info("{} Started Streams client", logPrefix);
+        } else {
+            // if transition failed, it means it is not in CREATED state
+            // any more; return immediately in this case
+            log.warn("{} Already stopped, cannot re-start", logPrefix);
```

Comment:
This line will only be triggered if `setState` returns false but not throwing an exception. So if `start()` is called after it is closed of as a second time, an exception will be thrown in `setState`. But I agree that this comment is a bit confusing. I will fix it.